### PR TITLE
Adds a thread-safe queue without duplicates of the same type and use it to prevent multiple seek actions

### PR DESCRIPTION
**Pull request by MrTheMake**
Supersedes https://github.com/ppy/osu-framework/pull/757

### CODE REVIEW

File: osu.Framework/Lists/SingleTypeQueue.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                    return false;
+
+                result = list[0].Element;
+                list.RemoveAt(0);
+
+                return true;
+            }
+        }
+
+        private class ListElement
```
Comment (RemieRichards) [inline osu.Framework/Lists/SingleTypeQueue.cs]: Should this not be a Struct?

Comment (MrTheMake) [inline osu.Framework/Lists/SingleTypeQueue.cs]: You are right.

**Comment by peppy** [comment]
Why is `elementType` an int? Can't we just use `GetType()` and `Type`s rather than int constants?

**Comment by peppy** [comment]
Closing this for the time being as the only case where this matters currently (seeking) has been resolved by other means (#1069).

**Pull request closed without merging**